ACKReactiveExtensions/Core/Codable.swift

```diff
@@ -0,0 +1,52 @@
+import ReactiveSwift
+
+/**
+ * Protocol that allows creation of custom Decoding errors
+ */
+public protocol DecodingErrorCreatable: Error {
+
+    /**
+     * Create error containing passed `DecodingError`
+     *
+     * - parameter decodeError: `DecodingError` which should be wrapped
+     */
+    static func createDecodeError(_ decodingError: DecodingError) -> Self
+}
+
+extension DecodingError: DecodingErrorCreatable {
+    public static func createDecodeError(_ decodingError: DecodingError) -> DecodingError {
+        decodingError
+    }
+}
+
+
+public extension SignalProducer where Value == Data, Error: DecodingErrorCreatable {
+    /// Decodes given `type` from received data using given `decoder`
+    ///
+    /// By default the `type` is inferred from return value
+    func decode<ResultType: Decodable>(type: ResultType.Type = ResultType.self, using decoder: JSONDecoder = ACKReactiveExtensionsConfiguration.jsonDecoder) -> SignalProducer<ResultType, Error> {
+        lift { $0.decode(type: type, using: decoder) }
+    }
+}
+
+public extension Signal where Value == Data, Error: DecodingErrorCreatable {
+    /// Decodes given `type` from received data using given `decoder`
+    ///
+    /// By default the `type` is inferred from return value, if no `decoder` is provided then `ACKReactiveExtensionsConfiguration.jsonDecoder` is used
+    func decode<ResultType: Decodable>(type: ResultType.Type = ResultType.self, using decoder: JSONDecoder = ACKReactiveExtensionsConfiguration.jsonDecoder) -> Signal<ResultType, Error> {
+        attemptMap { data -> Result<ResultType, Error> in
+            if ACKReactiveExtensionsConfiguration.allowMappingOnMainThread == false {
+                assert(Thread.current.isMainThread == false, "Mapping should not be performed on main thread!")
+            }
+            
+            do {
+                let decoded = try decoder.decode(ResultType.self, from: data)
+                return .success(decoded)
+            } catch {
+                // swiftlint:disable:next force_cast
+                let decodingError = error as! DecodingError
```

Comment:
Personally I don't like this force cast here. We are relying on the underlying implementation and Apple can change that in any future version.

I would prefer one `catch` for the `DecodingError` itself and then another `catch` the default one for the general error. What do you think?

Reply:
Well this is now solved by introducing generic `MappingError` that allows throwing `.mapping` error and `generic` error.

This generic enum is introduced so it can be used for Marshal extensions later on as it has the same problem. But I think that should not be part of this PR so it will be implemented some time in the future after this PR is merged.